Use per-derivation sandbox profiles

pikajude opened the pull request:
@copumpkin @henrytill @jwiegley @datakurre and anybody else interested in Darwin stuff.

This solution is in my opinion more viable long-term than continually adding more permissions to the default nix sandbox.

See https://github.com/NixOS/nix/pull/695

offlinehacker committed:
add support for auto format, format options

Commit:
start on sandbox stuff

Commit:
use per-derivation sandbox profiles

Commit:
remove useless muting lines; they interfere with debugging

datakurre commented:
As a user I'm positive for trying this out, but I recall @copumpkin being considered about some security implications.

ozanvos committed:
closure-compiler: 20150609 -> 20151015

Commit:
allow networking by default and remove it from derivations

Commit:
remove debugging from perl 5.16

zimbatm committed:
webfs: init at 1.21

offlinehacker committed:
systemd module: Add OnFailure to systemd common unit options

offlinehacker committed:
docker-gc: init at b0cc52aa3da2e2ac0080794e0be6e674b1f063fc

offlinehacker committed:
docker: 1.8.1 -> 1.9.0

offlinehacker committed:
docker: make lxc optional

bendlas committed:
wine: 1.7.53 -> 1.7.55

bendlas committed:
dropbox: 3.10.9 -> 3.10.11

peti committed:
configuration-hackage2nix.yaml: update list of broken builds

peti committed:
hackage-packages.nix: update Haskell package set

This update was generated by hackage2nix v20150922-44-g2042044 using the following inputs:

  - Nixpkgs: https://github.com/NixOS/nixpkgs/commit/8d34a1fc4bfcddbb1cc10d70958293e9e3b16298
  - Hackage: https://github.com/commercialhaskell/all-cabal-hashes/commit/770d6964c53cbf749adf3b8984b6a49700bb9db0
  - LTS Haskell: https://github.com/fpco/lts-haskell/commit/8ec1fc34193fe53e8e257219cb2e4618f82cade5
  - Stackage Nightly: https://github.com/fpco/stackage-nightly/commit/5236aaf5765a4a921ddc92b1e217e882b7b2f8cf

vizanto committed:
boot: 2.2.0 -> 2.4.2

lucabrunox committed:
gnome3_16: mark gnome-screenshot as broken

manveru committed:
google-fonts: init at 2015-11-18

ocharles committed:
devmon: Non-root user, set PATH, require udisks2

devmon refuses to run as root. Instead, we now run it as a user service,
and enable udisks2 in order to perform the mounts.

pSub committed:
Merge pull request #11121 from manveru/google-fonts

google-fonts: new package

peti committed:
r-modules: update list of broken builds to avoid Hydra errors

pSub committed:
Merge pull request #11120 from vizanto/pr-boot

boot: 2.2.0 -> 2.4.2

vcunat committed:
gnome-screenshot: unbreak build by 3.16.0 -> 3.16.2

/cc @lethalman.
Includes a revert of 1f27976e03c.

viric committed:
Updating freecad to 0.15

(cherry picked from commit 8a18bc8722dc1739b5b2109d556032ca9a068bbd)

Conflicts:
	pkgs/top-level/all-packages.nix

peti committed:
gnucash-2.4.15: disable test suite

The test suite was broken by the GLib 2.46.0 update. If anyone finds a
patch that remedies these issues for 2.4.x, please check it in!
Otherwise, we'll eventually update our default version to 2.6.x, I
suppose.

Closes https://github.com/NixOS/nixpkgs/issues/11084.

pSub committed:
calibre: 2.44.0 -> 2.44.1

peti committed:
nixos: configure bash prompt in Emacs' shell-mode, even if TERM=dumb

The idea that the interactive bash prompt isn't set in case of TERM=dumb
is intended to fix problems when other machines log remotely into a
NixOS installation via Tramp. A side-effect that change was, however,
that Emacs' shell-mode no longer had a correct prompt. I suppose the
presence of

  INSIDE_EMACS=24.5.2,comint

is a sufficiently unique indication that the current interactive shell
is running inside of an Emacs and that the prompt can thus be configured
safely.

peti committed:
nixos: remove debug code accidentally checked in with 833a5971568f8b5ea189126d61d8461d039bb93c

peti committed:
Merge pull request #11128 from et4te/source-code-pro-update

Updated source-code-pro SHA256

cillianderoiste committed:
Merge pull request #11034 from zimbatm/webfs-1.21

webfs: init at 1.21

Commit:
ipmitool: Fix build on darwin

s6_addr16 is not available on Mac OS X.

peti committed:
eukleides: switch to the new texlive implementation

peti committed:
asymptote: switch to the new texlive implementation

jagajaga committed:
Merge pull request #11116 from bendlas/update-dropbox-master

dropbox: 3.10.9 -> 3.10.11

joachifm committed:
electrum: 2.5.1 -> 2.5.4

offlinehacker committed:
Merge pull request #11065 from offlinehacker/pkgs/docker/optional_lxc

docker: make lxc optional

offlinehacker committed:
Merge pull request #11063 from offlinehacker/pkgs/docker/update_1.9.0

docker: 1.8.1 -> 1.9.0

offlinehacker committed:
Merge pull request #11057 from offlinehacker/pkgs/docker-gc/add

docker-gc: init at b0cc52aa3da2e2ac0080794e0be6e674b1f063fc

offlinehacker committed:
Merge pull request #11056 from offlinehacker/nixos/systemd/onfailure

systemd module: Add OnFailure to systemd common unit options

offlinehacker committed:
Merge pull request #10218 from offlinehacker/nixos/format_options

add support for auto format, format options

offlinehacker committed:
Merge pull request #11129 from et4te/font-awesome-sha-update

Updated SHA256 of font-awesome-ttf

offlinehacker committed:
Merge pull request #11131 from svend/ipmitool-darwin

ipmitool: Fix build on darwin

offlinehacker committed:
Merge pull request #11132 from joachifm/electrum-update

electrum: 2.5.1 -> 2.5.4

jagajaga committed:
slim service: update default theme

Jomik committed:
Add focus_password if defaultUser is set

Add 'focus_password yes' if defaultUser is set, to reflect the description of defaultUser.

abbradar committed:
Merge pull request #11114 from bendlas/update-wine

wine: 1.7.53 -> 1.7.55

wkennington committed:
btrfsProgs: 4.3 -> 4.3.1

oconnorr committed:
coq: 8.5b2 -> 8.5b3

roconnor committed:
Merge pull request #11140 from roconnor/master

coq: 8.5b2 -> 8.5b3

vbgl committed:
Merge pull request #11024 from osener/closure-compiler-20151015

closure-compiler: 20150609 -> 20151015

edolstra committed:
Merge pull request #11136 from Jomik/patch-1

Add focus_password if defaultUser is set

edolstra committed:
Revert "kernel: Remove unsupported 3.10, 3.12, 3.14"

This reverts commit 2441e002e26d60e62306ae03a2c0d42fe156f129. The
motivation for removing them was not very convincing. Also, we need
3.14 on some Hydra build machines.

edolstra committed:
services.cron.enable: Remove misleading default

Elsewhere we set it to "mkDefault (allFiles != [])" to the default has
no effect.

viric committed:
Fixing the evaluation broken at 8a18bc8722dc1739b5b2109d556

niksnut pointed this out in
https://github.com/NixOS/nixpkgs/commit/8a18bc8722dc1739b5b2109d556032ca9a068bbd#commitcomment-14498732

edolstra committed:
ceph: Remove old and git versions

edolstra committed:
ceph: Don't use symlinks

No need to invent an ad-hoc symlink scheme when we can handle this
kind of indirection at the Nix expression level (e.g. by doing
"ceph-dev = ceph").

annejan committed:
Updated to 1.0.5

annejan committed:
Update default.nix

cillianderoiste committed:
Merge pull request #11152 from annejan/patch-1

QtPass: updated to 1.0.5

cillianderoiste committed:
qtpass: tidy up

Commit:
Merge branch 'staging' of https://github.com/NixOS/nixpkgs into sandbox-profiles

nbp commented on pkgs/applications/version-management/git-and-tools/git/default.nix:
```diff
@@ -36,6 +36,9 @@ stdenv.mkDerivation {
   NIX_LDFLAGS = stdenv.lib.optionalString (!stdenv.isDarwin) "-lgcc_s";
 
+  # without this, git fails when trying to check for /etc/gitconfig existence
+  __propagatedSandboxProfile = stdenv.lib.sandbox.allowDirectoryList "/etc";
```
`__` prefixes are used for builtins.  I would recommend against this naming convention, and suggest that you use a single `_` prefix if such prefixes are needed.

shlevy commented:
Merged to staging, this PR will close at next staging merge.

shlevy commented:
Ah, didn't see @nbp's comment. Please fix and open a PR against staging when ready.

shlevy commented:
@pikajude Isn't generate-framework-profile inherently impure? Why is that here and not part of the build hook? It being here will mean the derivation will be different if instantiated on different versions.

pikajude commented:
@shlevy And it's specifically never substituted, which means that the derivation will have the same hash on host systems with different structures, but will always contain the dependencies that are specific to that system. I wanted to avoid a situation where people could have a copy of the profile substituted from Hydra that would cause link failures when building other packages.

I liked this solution better than using the pre-build-hook in Nix, though for all intents and purposes they're the same. Do you have a compelling argument for relegating the functionality back into Nix over keeping it here?

shlevy commented:
I'm not sure what you mean by "it's never substituted". Derivations will have a different `__sandboxProfile` string on different systems, right? That will result in a different hash.

pikajude commented:
No, how would it have changed? The definition of `generateFrameworkProfile` will be the same across systems, but the contents will vary.

shlevy commented:
Ah, so you use stuff like `__propagatedSandboxProfile = stdenv.lib.sandbox.importProfile (generateFrameworkProfile name);`. That's really broken, if the result of generateFrameworkProfile is substituted (or inherited from an old version of NixOS, etc.), it won't work. We can't have impurities like that in nixpkgs.

pikajude commented:
As I said, that's why I set `allowSubstitutes = false`.

shlevy commented:
There's a huge difference between an impure path in the store and an impurity in the system itself. If I have a build farm set up, how can I guarantee that the machine used to build the profile is the same as the one used to build the resultant derivation? I can't. If `pre-build-hook` needs to be improved, let's improve it, if we need some different *principled* way of ensuring this impurity is localized, let's implement it, but doing it this way breaks nix.

pikajude commented:
Ok, makes sense. The old `pre-build-hook` still exists and should be able to read __impureHostDeps, so should I just revert to that?

shlevy commented:
For now, yeah, I think so (though not on this branch, as it's already merged into staging :smile: ). Can you maybe open up an issue on nix for what you would like the interface/error messages/etc. coming from a better mechanism might look like?

pikajude commented:
@shlevy I can use `__propagatedImpureHostDeps` on frameworks only if you want, and keep sandbox profiles everywhere else.

shlevy commented:
Cool, thanks

pikajude commented:
@shlevy also @nbp proposed not using underscores at all for the sandbox profile attributes, how do you feel about that? It does make derivations look a bit cleaner.

shlevy commented:
Fine by me.

pikajude commented:
@shlevy: see https://github.com/NixOS/nixpkgs/pull/11193

et4te committed:
Updated source-code-pro SHA256.

et4te committed:
Updated SHA256 of font-awesome-ttf

pikajude commented:
@edolstra btw, urllib3 DOES require network in its checkPhase, do you have any idea why it passes on Linux?

shlevy commented:
@pikajude @edolstra I think Linux chroot enables localhost networking, right? Is that all urllib3 needs?

copumpkin commented:
The OSX sandbox profile we were using enabled localhost networking too, so
that shouldn't break.

On Mon, Nov 23, 2015 at 9:29 PM, Shea Levy <notifications@github.com> wrote:

> @pikajude <https://github.com/pikajude> @edolstra
> <https://github.com/edolstra> I think Linux chroot enables localhost
> networking, right? Is that all urllib3 needs?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/11001#issuecomment-159054846>.
>

ncfavier commented:
Does anyone remember why `/dev/ptmx` was needed for vim, and if it can be removed now?
